test/distributed/tensor/parallel/test_fsdp_2d_parallel.py

```diff
@@ -215,21 +217,56 @@ def _test_2d_e2e_flow(
                 print(name, param_names_2d)
             self.assertTrue(name in param_names_2d)
         self._compare_params(model, model_2d)
-
         if multi_param_group and use_orig_params:
             param_group = [
                 {"params": model.net1.parameters(), "lr": 0.02},
                 {"params": model.net2.parameters(), "lr": 0.15},
             ]
-            optim = torch.optim.Adam(param_group, lr=0.01)
+            if optim_in_backward:
+                _apply_optimizer_in_backward(
+                    optimizer_class=torch.optim.Adam,
+                    params=param_group[0]["params"],
+                    optimizer_kwargs={"lr": param_group[0]["lr"]},
+                )
+                _apply_optimizer_in_backward(
+                    optimizer_class=torch.optim.Adam,
+                    params=param_group[1]["params"],
+                    optimizer_kwargs={"lr": param_group[1]["lr"]},
+                )
+            else:
+                optim = torch.optim.Adam(param_group, lr=0.01)
             param_group = [
                 {"params": model_2d.net1.parameters(), "lr": 0.02},
                 {"params": model_2d.net2.parameters(), "lr": 0.15},
             ]
-            optim_2d = torch.optim.Adam(param_group, lr=0.01)
+            if optim_in_backward:
+                _apply_optimizer_in_backward(
+                    optimizer_class=torch.optim.Adam,
+                    params=param_group[0]["params"],
+                    optimizer_kwargs={"lr": param_group[0]["lr"]},
+                )
+                _apply_optimizer_in_backward(
+                    optimizer_class=torch.optim.Adam,
+                    params=param_group[1]["params"],
+                    optimizer_kwargs={"lr": param_group[1]["lr"]},
+                )
+            else:
+                optim_2d = torch.optim.Adam(param_group, lr=0.01)
         else:
-            optim = torch.optim.Adam(model.parameters(), lr=0.01)
-            optim_2d = torch.optim.Adam(model_2d.parameters(), lr=0.01)
+            if optim_in_backward:
+                _apply_optimizer_in_backward(
```

Comment:
Does this also work with DDP? If so, do you mind also testing with DDP. It can be in a separate PR. Thanks!

Reply:
`_apply_optimizer_in_backward` works for DDP too.